### PR TITLE
feat: debian datasource

### LINKED ISSUE #7041: Add apt datasource for private repositories
**What would you like Renovate to be able to do?**

https://github.com/renovatebot/renovate/issues/3722 was closed because we can manage packages pining with public apt repositories throug the repology datasource. It does not work for private repositories and any repository that is not indexed by repology.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

I create Docker layers with the apt packages installed in it to ensure immutability through the Docker datasource. But that is not an optimized solution at all.

### PR DESCRIPTION
**Pull request by Ka0o0**
## Changes:

- Adds a Debian package datasource (Closes #7041)

## Context:

This datasource is able to search multiple Debian package sources that provide a Packages file as described [here](https://wiki.debian.org/DebianRepository).
This repository standard defines components within a release repository (e.g. main within bullseye release).

The datasource works by downloading the compressed Packages file for each component and extracting it. Currently, only .xz files are supported, but it should be possible to add other compression standards as well.
The file is downloaded using curl, which is a shell cmd call, into a configurable folder.
The file is only downloaded if the local version is older than what the server has (using curl's  -z parameter) and is only extracted if needed.

The file is then searched through, currently, line by line which is a bit slow with Node.JS and further work might improve the performance by removing unused information from the extracted files.
It then returns the latest version which is, differently from what repology is returning us, the version you can specify with apt install, like `apt-get install curl=7.74.0-1.3+deb11u1`. An example PR created with this banch merged looks like this: https://github.com/Ka0o0/renovate-testrepo/pull/2

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

## Open Points

- [ ] Where to add documentation?
- [ ] Is the current implementation regarding the datasource configuration valid like it is currently? 
- [ ] This datasource calls some third party tools which are currently available in the renovate Docker image. Are any changes to how this is used required, like explicitly stating the dependency somewhere?

### CODE REVIEW

**Comment by HonkingGoose** [comment]
>  - [ ] Where to add documentation?

Once the code for the feature is approved by the maintainers, the documentation can be added in a new `lib/datasource/deb/readme.md` file, I think.

The "Repology datasource" has docs that you can look at for inspiration. [^repology-datasource] 

[^repology-datasource]: https://github.com/renovatebot/renovate/blob/main/lib/datasource/repology/readme.md

**Comment by rarkins** [comment]
Hi @Ka0o0, thanks for this PR! Can you create a feature request issue to link to from this PR, and so we can discuss the requirements/implementation proposal there first?

**Comment by Ka0o0** [comment]
Hi @rarkins,

thanks for pointing that out. Actually, there is #7041 already opened which I've referenced in this PR's description.

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [inline lib/datasource/deb/index.spec.ts:3]: Tests need to mock any web requests, so you should try to create a minimal packages file

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:41]: nope, use `ensureCacheDir` from `util/fs`, checkout `terraform manager for sample usage

https://github.com/renovatebot/renovate/blob/4f65b57225a42f9deb010f8dbfeb06eb94c0dd69/lib/manager/terraform/lockfile/hash.ts#L93-L102

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:104]: nope, use `this.http.stream` to download files, see terraform manager (check the lockfile dir)

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:128]: We should use a javascipt based extraction, see terraform manager

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/datasource/deb/index.spec.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.spec.ts:3]: That makes sense to me. Is it okay to store a compressed file within this repository and serve it with nock's `replyWithFile` parameter? Or should the content be "clear" text within the source code and be compressed on the fly?

Comment (viceice) [reply lib/datasource/deb/index.spec.ts:3]: Can be a compressed file, but should be as small as possible. But you can also compress on the fly if it's fast.

File: lib/datasource/deb/index.ts
Comment (viceice) [reply lib/datasource/deb/index.ts:34]: Yes, it's validates at some places or validation is planned.

File: lib/datasource/api.ts
Comment (HonkingGoose) [inline lib/datasource/api.ts:6]: If you want, you can add yourself to the list of `contributors` in the `package.json` file. 😉 

https://github.com/renovatebot/renovate/blob/cc016ac14c3aa3183825b16e67a40b983def9adf/package.json#L78-L116

File: lib/datasource/deb/index.spec.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.spec.ts]: Done

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts]: Done

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts]: Done as well

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts]: Done

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [inline lib/datasource/deb/index.ts:71]: add a comment

**Comment by viceice** [comment]
Please do not use simple `Done` like comments, they produce a lot of noise.

@HonkingGoose Can we extend our contribution guide? So people should simply resolve conversations when done and only use re-request review button. Any only if no maintainer answer, use comments after ~5days.

@rarkins WDYT?

**Comment by HonkingGoose** [comment]
> @HonkingGoose Can we extend our contribution guide? So people should simply resolve conversations when done and only use re-request review button. Any only if no maintainer answer, use comments after ~5days.

Sounds good to me. Is the goal here to reduce noise, and to "teach" contributors to wait a bit before commenting?

Our `.github/contributing.md` file says: [^quote]

> ## Re-requesting a review
> 
> Please do not ping your reviewer(s) by mentioning them in a new comment.
> Instead, use the re-request review functionality.
> Read more about this in the [GitHub docs, Re-requesting a review](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#re-requesting-a-review).

[^quote]: https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [inline lib/datasource/deb/index.spec.ts:2]: use `fs-extra` and use async.

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:89]:
```suggestion
```

unused

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:1]:
```suggestion
```
unused

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:210]:
```suggestion
      input: fs.createReadStream(extractedFile),
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:4]:
```suggestion
```

Use our `fs` layer, add missing functions if required

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:7]:
```suggestion
```

use our `fs.pipeline`

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:8]:
```suggestion
```
unused

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:10]:
```suggestion
import type { GetReleasesConfig, ReleaseResult } from '..';
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:13]:
```suggestion
import type { HttpOptions } from '../../util/http';
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:15]:
```suggestion
import type { DebLanguageConfig, PackageDescription } from './types';
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:5]: What is this for?

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:311]:
```suggestion
```

not needed, no benefit

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:317]:
```suggestion
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:54]: This should be hard coded, no need to be configurable

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:52]: Why now add this to registry url too. Make it optional and fallback to `amd64` by default?

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:115]: Can be much simplier:
```suggestion
    let extractedFileExists = fs.pathExists(extractedFile);
```

Checkout `util/fs/proxies` 😉 
https://github.com/renovatebot/renovate/blob/c5d50f6da3502d9f2a8f805fb65292a98246a6bf/lib/util/fs/proxies.ts#L60-L63

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:100]:
```suggestion
    const hashedPackageUrl = hash.digest('hex');
```

No need for copy

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:128]:
```suggestion
          await this.extract(compressedFile, compression, extractedFile);
          await fs.rm(compressedFile);
```

delete compressed file after extraction

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:223]: Are you sure we only ever have one version?

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:190]: This is a goot candidate to add cache decorator, sample:

https://github.com/renovatebot/renovate/blob/4f65b57225a42f9deb010f8dbfeb06eb94c0dd69/lib/manager/terraform/lockfile/hash.ts#L61-L69

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts:52]: good idea, I'll then rename the current binary arch configuration to "defaultBinaryArch"

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts:223]: So yes and no, every Packages file must only have a package once but a package might be specified in multiple repositories which I've added now.

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts:135]: Actually, I think can use the timestamp of the extracted file, which should be roughly the same as the downloaded file.

Comment (viceice) [reply lib/datasource/deb/index.ts:135]: I don't think so, it depends on extraction code. It can use embedded metadata in compressed file and set the file timestamp, so it will always older that the compressed file date.

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts:184]: Mhm but what if the server returns a new compressed file, can I manually invalidate the cache entry then?

**Comment by Ka0o0** [comment]
Hi @viceice , 
I saw that you have referenced this PR. Is there anything I can do to move forward with this PR?

**Comment by viceice** [comment]
Fix coverage issues?

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [inline lib/datasource/deb/index.spec.ts:11]: We have helpers for this, checkout fixtures class

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [inline lib/datasource/deb/index.spec.ts:154]:
```suggestion
          expect(res?.releases).toHaveLength(2);
```

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:70]: Make static

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:72]: Make static when no this is used

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:94]: Use `upath.join`

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:106]: See above

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:128]: Please use logger metadata object instead of string concat.

File: lib/datasource/deb/index.ts
Comment (viceice) [inline lib/datasource/deb/index.ts:185]: I think we should cache the compressed file and directly stream if possible. Would sava a lot disk space. 🤔

**Review by viceice — COMMENTED** [review]
@rarkins I'm still somehow unconfident with the current cache mechanism. I think we should extract all updates and write to renovate cache with additional metadata. And only use cache if still valid.

File: lib/datasource/deb/index.spec.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.spec.ts:57]: Unfulfilled mocks are only evaluated after the test block. In the next line the array access on `getTrace()` would therefore be executed which would lead to an runtime exception in a failing test case scenario.

File: lib/datasource/deb/index.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.ts:197]: I agree that would indeed some disk space for the cost of computational expense. I was also thinking about removing some unused meta information from the extracted Package file which would also reduce the package file size and reduce the lookup time (which is more important I guess). What do you think?

File: lib/datasource/deb/index.spec.ts
Comment (Ka0o0) [reply lib/datasource/deb/index.spec.ts]: The only test case where I think a fixture would make sense is [this one](https://github.com/Ka0o0/renovate/blob/feat/deb-datasource/lib/datasource/deb/index.spec.ts#L44). However, I don't quite know how to get it running.

If I understood it correctly, the Fixtures.mock method will register some temp files based on what is provided as a parameter and intercept FS calls to those files? So I'm setting up the test case like this (basically replacing line 44 - 45):

```JavaScript
const fixtures: DirectoryJSON = {};
fixtures[extractedPackageFile] = Fixtures.get('Packages');
Fixtures.mock(fixtures);
```
But now the test case would fail [here](https://github.com/Ka0o0/renovate/blob/feat/deb-datasource/lib/datasource/deb/index.spec.ts#L46). Shouldn't the Fixtures' mock make sure that the file exists?

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [reply lib/datasource/deb/index.spec.ts:59]: Yes and no, if you look at the http mock code, you can see a `afterEach` which will validate this, so we don't have to repeat it here. 😉

**Comment by rarkins** [comment]
Needs test fixes

File: index.mjs
Comment (viceice) [inline index.mjs:1]: why this file?

File: lib/util/fs/index.ts
Comment (viceice) [inline lib/util/fs/index.ts:163]: move to `./proxies.ts`

File: lib/datasource/deb/index.spec.ts
Comment (viceice) [inline lib/datasource/deb/index.spec.ts:1]: file structure changed, please move datasource to right folder.

please don't force-push

**Comment by rarkins** [comment]
Closing due to staleness

**Pull request closed without merging**
